### PR TITLE
238. Product of Array Except Self

### PR DESCRIPTION
**Pull request by t-ooka**
[問題リンク](https://leetcode.com/problems/longest-consecutive-sequence/description/)

## 問題文
Given an integer array nums, return an array answer such that answer[i] is equal to the product of all the elements of nums except nums[i].

The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.

You must write an algorithm that runs in O(n) time and without using the division operation.

 

Example 1:
```
Input: nums = [1,2,3,4]
Output: [24,12,8,6]
```

Example 2:
```
Input: nums = [-1,1,0,-3,3]
Output: [0,0,9,0,0]
```

Constraints:

- 2 <= nums.length <= 105
- -30 <= nums[i] <= 30
- The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.

### CODE REVIEW

File: Product of Array Except Self/step2-using-LR-lists.cpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+            L[i] = L[i-1] * nums[i-1];
+        }
+
+        vector<int> R(length);
+        R[length-1] = 1;
+        for (int i = length-2; i >= 0; i--) {
+            R[i] = R[i+1] * nums[i+1]; 
+        }
+
+        vector<int> ans(length);
```
Comment (nodchip) [inline Product of Array Except Self/step2-using-LR-lists.cpp:18]: 変数名には原則省略形は用いないことをお勧めいたします。
https://google.github.io/styleguide/cppguide.html#General_Naming_Rules
> Minimize the use of abbreviations that would likely be unknown to someone outside your project (especially acronyms and initialisms). Do not abbreviate by deleting letters within a word. As a rule of thumb, an abbreviation is probably OK if it's listed in Wikipedia. Generally speaking, descriptiveness should be proportional to the name's scope of visibility.

File: Product of Array Except Self/note.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+```sample.py
+input = [1,2,3,4]
+
+left = [1,2,6,24]
+right = [24,12,4,1]
+```
+
+left,rightはそれぞれ、各々のインデックスの数までの積を表しているので、両隣の数の積を取れば自身をのぞいた積にになる。
+
+- 時間計算量 o(n)
```
[the remark this answers is not in the document]
Comment (t-ooka) [reply Product of Array Except Self/note.md:15]: 了解しました。以後Oで統一します。

File: Product of Array Except Self/step2-using-LR-lists.cpp (hunk 1 above)
Comment (t-ooka) [reply Product of Array Except Self/step2-using-LR-lists.cpp:18]: わかりました。省略形は極力避けるようにいたします。

File: Product of Array Except Self/note.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+suffix = 1
+for i in range(len(nums)-1, -1, -1):
+  res[i] *= suffix
+  suffix *= nums[i]
+
+return res
+```
+
+-時間計算量 o(n)
+-空間計算量 o(1)
```
[the remark this answers is not in the document]
Comment (t-ooka) [reply Product of Array Except Self/note.md:40]: レスポンスとして使用する値については空間計算量に含まない認識で書いておりました。
ソースコードベースですとおっしゃるとおり、O(n)です、、！

**Comment by t-ooka** [comment]
数週間空けて解けたのでマージします。

**Pull request merged**